server/echoserver.go

```diff
@@ -1,39 +1,72 @@
 package main
 
 import (
+	"crypto/rand"
+	"encoding/hex"
 	"fmt"
 	"io/ioutil"
 	"log"
 	"net"
 	"net/http"
 	"os"
-	
+
 	"gopkg.in/alecthomas/kingpin.v2"
 )
 
+var authToken string
+
 func echo(w http.ResponseWriter, r *http.Request) {
 	var response string
 	out, err := ioutil.ReadAll(r.Body)
 	response = string(out)
 	if err != nil {
-		response = fmt.Sprintf("there was error: " + err.Error())
+		response = fmt.Sprintf("there was error reading request: " + err.Error())
 	}
 	fmt.Fprintf(w, response)
 }
 
+func isAuthorized(endpoint func(http.ResponseWriter, *http.Request)) func(http.ResponseWriter, *http.Request) {
```

Comment:
Nice work on this.  This signature alone is showing strong go skills compared to a month ago.